Add reset and clean for submodules

XavierChapron opened the pull request:
Related to https://github.com/actions/checkout/issues/358

danra commented on dist/index.js:
```diff
@@ -7416,6 +7436,8 @@ function getSource(settings) {
                     yield git.submoduleSync(settings.nestedSubmodules);
                     yield git.submoduleUpdate(settings.fetchDepth, settings.nestedSubmodules);
+                    yield git.submoduleReset(settings.nestedSubmodules);
+                    yield git.submoduleClean(settings.nestedSubmodules);
```
Shouldn't Reset and Clean be called prior to Update? I think the current code will fail if what was previously a folder belonging to the current repo is now the location of a submodule

danra replied:
Also, it seems like this cleaning should only occur if clean=true

XavierChapron replied:
Good point for the `clean=true`!
I am not sure about the importance of the order but you are probably right and I believe it would be more coherent with what is done for the main git repository. 

However, this project seems quite dead, and I will therefore not spend more time on this before getting the assurance that this could really be merged...

danra replied:
It's sad if it's dead, `actions/checkout@v2` is used probably in 99.99% of the git actions out there...

XavierChapron replied:
I do agree and I'm also using it a lot.
But no response on this PR, nor any commit on the project since 3 month. That's much for a project that is so necessary !

XavierChapron replied:
Should be done in the commits proposed by @AceCoderLaura that are added to the PR.

jeppefrandsen commented:
Any news on this issue? Would be nice to get fixed 🍻

AceCoderLaura commented:
I might pick up this PR and finish it. This is causing disruptions for my company.

AceCoderLaura commented:
This PR appears to break two tests, I will attempt to remedy this.

AceCoderLaura commented:
I think I've addressed the review items as well as passed the tests. Please fast forward this branch to https://github.com/AceCoderLaura/checkout/tree/finishing-628 if my changes are satisfactory.

XavierChapron commented:
> I think I've addressed the review items as well as passed the tests. Please fast forward this branch to https://github.com/AceCoderLaura/checkout/tree/finishing-628 if my changes are satisfactory.

Hello, thank you!
You have many whitespace change in your commit https://github.com/AceCoderLaura/checkout/commit/8328374403a422ebe7803250ea3ced9d8f07803d
Can you change this? Or if the change are legetimate, maybe create a specific commit?

Then I will be pleased to cherry-pick your commits. And I hope someone will review and merge this anytime soon!

AceCoderLaura commented:
Ah yes, my git configuration wasn't setup to convert line endings back to UNIX-style. I'll fix it some time today if I get some time.

AceCoderLaura commented:
I've pushed a version with only the changes to the TypeScript source files here: https://github.com/AceCoderLaura/checkout/tree/finishing-628-no-compile I will have to ask you to build the project yourself because I've run out of time to fix the line endings issue. Hope this helps!

AceCoderLaura commented:
I got some more time to work on this today and successfully transpiled the change to index.js without any disruptions to the line endings.

XavierChapron commented:
> I got some more time to work on this today and successfully transpiled the change to index.js without any disruptions to the line endings.

Thanks a lot, I've included your commit in this PR.

AceCoderLaura commented:
We're still using this fork but it's getting stale and it now has conflicts with the main branch. I haven't tried the latest version, does anyone know if this is still an issue in the main branch? If it is, I'll make some time to update the forked version.

OscarVanL commented:
I ran into this issue recently. For me, a file in a submodule got modified by a previous run and then was present in the next run, causing issues.

It would be nice if this got merged so I can have reliable CI without adding hacks.

This was my workaround by the way:
```
      - uses: actions/checkout@v3
        with:
          submodules: true
      - name: Submodule cleanup fix  # Bodge for https://github.com/actions/checkout/issues/358
        run: |
          git submodule foreach --recursive git clean -ffdx
          git submodule foreach --recursive git reset --hard
```

Hish15 commented:
What is preventing this from being merged ?
